ckan/tests/controllers/test_package.py

```diff
@@ -1752,6 +1799,115 @@ def test_search_with_extra_params(self, app, monkeypatch):
+            ".dataset-list " ".dataset-item " ".dataset-heading a"
+        )
+        ds_titles = [n.string for n in ds_titles]
+
+        assert len(ds_titles) == 1
+        assert "Dataset One" in ds_titles
+        assert "Dataset Two" not in ds_titles
+
+    @pytest.mark.ckan_config("ckan.search.show_all_types", "false")
+    def test_search_page_respects_show_all_types_param_false(self, app):
+        org = factories.Organization()
+
+        factories.Dataset(
+            name="dataset-one",
+            title="Dataset One",
+            type="dataset",
+        )
+        factories.Dataset(
+            name="dataset-two",
+            title="Dataset Two",
+            type="fancy_type",
+        )
+
+        search_url = url_for("dataset.search")
+        search_response = app.get(search_url)
+
+        search_response_html = BeautifulSoup(search_response.data)
+        ds_titles = search_response_html.select(
+            ".dataset-list " ".dataset-item " ".dataset-heading a"
+        )
+        ds_titles = [n.string for n in ds_titles]
+
+        assert len(ds_titles) == 1
+        assert "Dataset One" in ds_titles
+        assert "Dataset Two" not in ds_titles
+
+    @pytest.mark.ckan_config("ckan.search.show_all_types", "true")
+    def test_search_page_respects_show_all_types_param_true(self, app):
+        org = factories.Organization()
+
+        factories.Dataset(
+            name="dataset-one",
+            title="Dataset One",
+            type="dataset",
+        )
+        factories.Dataset(
+            name="dataset-two",
+            title="Dataset Two",
+            type="fancy_type",
+        )
+
+        search_url = url_for("dataset.search")
+        search_response = app.get(search_url)
+
+        search_response_html = BeautifulSoup(search_response.data)
+        ds_titles = search_response_html.select(
+            ".dataset-list " ".dataset-item " ".dataset-heading a"
+        )
+        ds_titles = [n.string for n in ds_titles]
+
+        assert len(ds_titles) == 2
+        assert "Dataset One" in ds_titles
+        assert "Dataset Two" in ds_titles
+
+    @pytest.mark.ckan_config("ckan.search.show_all_types", "fancy_type")
+    def test_search_page_respects_show_all_types_param_string(self, app):
+        org = factories.Organization()
+
+        factories.Dataset(
+            name="dataset-one",
+            title="Dataset One",
+            type="dataset",
+        )
+        factories.Dataset(
+            name="dataset-two",
+            title="Dataset Two",
+            type="fancy_type",
+        )
+
+        search_url = url_for("dataset.search")
+        search_response = app.get(search_url)
+
+        search_response_html = BeautifulSoup(search_response.data)
+        ds_titles = search_response_html.select(
+            ".dataset-list " ".dataset-item " ".dataset-heading a"
+        )
+        ds_titles = [n.string for n in ds_titles]
+
+        assert False
```

Comment:
TODO: what should the assertion be here?
I read https://docs.ckan.org/en/2.8/maintaining/configuration.html#ckan-search-show-all-types but I still don't really understand what the desired behaviour is in this situation, or whether the code as it stands is doing it.

Reply:
These docs are terrible. I think what they are trying to say is that when this config setting is "fancy_type" if you go to the "/fancy_type" page (custom dataset type page like /harvest or /showcase) you should get all dataset types listed.

Reply:
OK. That isn't what happens now if I set `ckan.search.show_all_types = "fancy_type"` I'm also now unclear on what `/fancy_type` should show if `ckan.search.show_all_types = False` and what should be there if `ckan.search.show_all_types = True` ?

Lets ignore what the code does at the moment and run over each case to work out what I'm aiming for. I think probably the **intention** is:

* `ckan.search.show_all_types = False` (or not set):
    * `/datasets`: show only `type=dataset`
    * `/fancy_type`: show only `type=fancy_type`
* `ckan.search.show_all_types = True`:
    * `/datasets`: show both `type=dataset` and `type=fancy_type`
    * `/fancy_type`: show only `type=fancy_type` (or should this also show both `type=dataset` and `type=fancy_type` in this case?)
* `ckan.search.show_all_types = "fancy_type"`:
    * `/datasets`: show only `type=dataset`
    * `/fancy_type`: show both `type=dataset` and `type=fancy_type`

does that seem like what we should be aiming for, or am I wrong?